frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -60,15 +84,48 @@ impl From<FeasibilityError> for MinerError {
 	}
 }
 
+/// Save a given call into OCW storage.
+fn save_solution<T: Config>(call: &Call<T>) {
+	let storage = StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL);
+	storage.set(&call);
```

Comment:
You should use the atomic `::mutate` here. https://github.com/paritytech/substrate/pull/7976/files

Reply:
Why? We don't care about the current value, so what benefit is there in reading it? We're not using the `get-check-set` pattern mentioned in [the docs](https://substrate.dev/rustdocs/v3.0.0/sp_runtime/offchain/storage/struct.StorageValueRef.html#method.set).

Reply:
https://github.com/paritytech/substrate/pull/7976#discussion_r563781298

Reply:
It's very surprising that `set` would allow for reads of an incomplete write, even from a separate, parallel worker. I traced the function calls back a bit, and ended up in [`trait sp_core::offchain::Externalities`](https://substrate.dev/rustdocs/v3.0.0/sp_core/offchain/trait.Externalities.html#tymethod.local_storage_set). For that trait, `local_storage_set` requires `&mut self`.

Obviously, we can OCWs in parallel, so the storage needs to be shared. I haven't traced it back exhaustively, but what I'd expect to find is something like `RwLock<dyn Externalities>`. Given that, an OCW which wants to read the storage has to wait until its peer with the write lock is completely finished writing, before it has a chance to read.

I believe that the atomic nature of the `mutate` method is to avoid this scenario:

- OCW A: read storage `Foo`
- OCW B: update storage `Foo`
- OCW A: update storage `Foo` based on its then-current value, clobbering what `B` wrote

AFAICT, normal mutable aliasing rules prevent OCW B from ever reading while OCW A is in the middle of writing. If I'm right, then we really don't need `mutate` here.

Reply:
I am totally happy to accept what you are proposing, if you _actually_ find out that it is designed and works as you hypothesize, not based on numerous assumptions that you are making. 

I am not sure about any of this myself, to be frank, and generally prefer consulting and the trusting the opinion of someone who actually does (Tomek in the case of the linked discussion). You can probably ping him if you want to clarify this.

Reply:
update? 

(learning from the past, I think you should just change this 1 line to a 1 liner that uses mutate and call it a day and be safe and save time as well.)

Reply:
Yes ok. I just can't find the actual implementation of offchain storage anywhere; everything seems to be abstractions that layer on abstractions. I'll just use `mutate here`. 8b31773fd